zconf/chp/lschp.8
.\" Copyright 2017 IBM Corp.
.\" s390-tools is free software; you can redistribute it and/or modify
.\" it under the terms of the MIT license. See LICENSE for details.
.\"
.TH LSCHP 8 "Mar 2007" s390\-tools

.SH NAME
lschp \- list information about available channel\-paths.

.SH SYNOPSIS
.B lschp
.RB [ \-h|\-\-help ]
.RB [ \-v|\-\-version ]
.RB [ \-\-format
.IR FORMAT ]
.RB [
.I CHPID
]

.SH DESCRIPTION
The lschp command lists status and type information about available
channel\-paths. A
.I CHPID
argument can be specified to return the information for the respective
channel-path identifier only.
Channel\-path identifiers are specified in hexadecimal notation either simply
as the CHPID\-number (e.g. e0) or in the form

.RS
<cssid>.<id>
.RE

where <cssid> is the channel\-subsystem identifier and <id> is the CHPID\-number (e.g. 0.7e).

.B Column description:

CHPID
.RS
Channel\-path identifier.
.RE

Vary
.RS
Logical channel\-path state:
.br
0 = channel\-path is not used for I/O
.br
1 = channel\-path is used for I/O
.RE

Cfg.
.RS
Channel\-path configure state:
.br
0 = stand\-by
.br
1 = configured
.br
2 = reserved
.br
3 = not recognized
.RE

Type
.RS
Channel\-path type identifier.
.RE

Cmg
.RS
Channel measurement group identifier.
.RE

Shared
.RS
Indicates whether a channel\-path is shared between LPARs:
.br
0 = channel\-path is not shared
.br
1 = channel\-path is shared
.RE

PCHID
.RS
Physical channel-ID unless the 4-digit hexadecimal value is enclosed in
parenthesis.

If the value is enclosed in parenthesis, no physical channel-ID is
associated with the CHPID, and the value is an internal channel-ID.

For example, 0501 specifies a PCHID whereas (0502) specifies an internal
channel-ID.
.RE

A column value of '\-' indicates that a facility associated with the respective
channel\-path attribute is not available.

.SH OPTIONS
.B \-h
.br
.B \-\-help
.RS
Print a short help text, then exit.
.RE

.B \-v
.br
.B \-\-version
.RS
Print version number, then exit.
.RE

.br
.B \-\-format \fIFORMAT\fR
.RS
Specify the output format. Supported formats include
\fIjson\fR, \fIjson-seq\fR, \fIjsonl\fR, \fIpairs\fR and \fIcsv\fR.
See section "OUTPUT FORMAT" for more details.
.RE


.SH "OUTPUT FORMAT"
This section contains additional information for json output format.

.SS json

JSON output consists of a top-level object with the following properties
(key-value pairs):

.IP \(bu 3
.BR meta :
Tool meta-data including API level, version, host name, and time of invocation
.PP
.IP \(bu 3
.BR channel-paths :
Information about available Channel\-path
.PP

Example JSON output for channel-path with all properties:

.nf
{
  "meta": {
    "api_level": 1,
    "version": "2.39.0-build-20260119",
    "host": "localhost",
    "time_epoch": 1769092556,
    "time": "2026-01-22 15:35:56+0100"
  },
  "channel_paths": [
    {
      "chpid": "0.32",
      "vary": 1,
      "cfg": 1,
      "type": 27,
      "cmg": 2,
      "shared": 1,
      "chid": 576,
      "chid_external": true
    },
    {
      "chpid": "0.34",
      "vary": 1,
      "cfg": 0,
      "type": 27,
      "cmg": null,
      "shared": null,
      "chid": 608,
      "chid_external": true
    },
    {
      "chpid": "0.f8",
      "vary": 1,
      "cfg": 1,
      "type": 36,
      "cmg": 3,
      "shared": 1,
      "chid": 2040,
      "chid_external": false
    }
  ]
}
.fi

.SH SEE ALSO
.BR chchp (8)

More information on the \fBChannel\-path type identifier\fR can be found at
.br
\fBhttps://www.ibm.com/docs/en/zos/3.1.0?topic=iee196i\-iee174i\-form\-2\-37\fR


CHANGELOG.md
Release history for s390-tools (MIT version)
--------------------------------------------

* __v2.xx.x (20xx-xx-xx)__

  For Linux kernel version: 6.x

  Add new tools / libraries:

  Changes of existing tools:

  Bug Fixes:

* __v2.41.0 (2026-02-16)__

  For Linux kernel version: 6.19

  Changes of existing tools:
  - chreipl: Make --bootparms work for ECKD re-IPL
  - cpacfstats: Add 'unauthorized' state to CPU-MF counters
  - cpictl: Detect RHCOS using VARIANT_ID
  - hsci: Automatically set appropriate MTU for HSCI
  - libutil: Add util_readlink() and util_readlinkat() helpers
  - libutil: Add util_startswith() to util_str
  - libutil: Add utility parsing functions
  - lschp: Add support for structured output (--format)
  - lsreipl: Suppress 'clear' output if not supported
  - pvimg: Add '--format text' support to 'pvimg info'
  - pvimg: Add '--print-schema <FORMAT>' option to 'pvimg info'
  - pvimg: Add '--show-secrets' flag to 'pvimg info'
  - pvimg: Provide improved JSON output to 'pvimg info --format json'
  - pvinfo: Improve User experience on non-SE enabled systems
  - zipl/ngdump: Ensure ext4 file system is used on dump partition
  - zkey: Add support for integrity protected disks using HMAC keys

  Bug Fixes:
  - cpumf/pai: Handle different size of perf_event_attr
  - lscss: Fix memory leak
  - zipl: Fix dump job on tape devices

* __v2.40.0 (2025-12-11)__

  For Linux kernel version: 6.18

  Add new tools / libraries:
  - Add project-wide .clang-tidy configuration
  - libutil: Introduce util_time for time related functionality
  - libutil: Introduce zsh/bash autocompletion tooling based on util_opt
  - pvinfo: Tool to display Secure Execution system information
  - pvverify: Tool to verify host-key documents

  Changes of existing tools:
  - cpumf: Implement zsh and bash autocompletion
  - dasdfmt: Implement zsh and bash autocompletion
  - dbginfo.sh: Add NetworkManager and netplan
  - dbginfo.sh: Add kvm_stat
  - dbginfo.sh: Adding stp time information
  - dbginfo.sh: Simplify procfs collection
  - hyptop: Add physical information row
  - hyptop: Calculate sample time delta for physical partition
  - hyptop: Replace long option names using _ with - for consistency
            For example: --cpu_types -> --cpu-types
            (Options with _ are still supported for backward compatibility)
  - libekmfweb: Add function to validate a certificate against the identity key
  - netboot: Add longer kernel command lines support
  - udev/rules.d: Make virtio-blk devices non-rotational
  - udev/rules.d: Set default io scheduler to 'none' for virtio-blk
  - ziomon: Add support to sample device symlinks (/dev/disk/...)
  - ziorep_config: Add fcp-lun details to -M option output
  - ziorep_config: Add port_id and failed attributes to -A option output
  - netboot: Install on non-s390 architectures

  Bug Fixes:
  - lib(ekmfweb|kmipclient): Use ln without -r
  - s390-tools: Fix various compilation issues with musl libc
  - zipl/boot: Fix unused loadparm when SCLP line-mode console is absent

* __v2.39.0 (2025-10-14)__

  For Linux kernel version: 6.16 / 6.17

  Changes of existing tools:
  - chpstat: Add options to select IEC units for scaling (SI units are default)
  - chzdev: Introduce --no-module-load option
  - cpi: Disable CPI for SEL guests by default
  - dbginfo.sh: Enhance logging on timeout triggered
  - iucvterm: Install symlink for lsiucvallow.8 man page
  - lshwc: Add command line flag to specify individual counters
  - lspai: Add command line flag for delta values
  - lspai: Add command line flag for short counter names
  - lspai: Add command line flag to specify individual counters
  - lspai: Add command line flags for all cpus
  - lspai: Add command line flags for hexadecimal output
  - man: Use CR for constant width font
  - pvimg: Add '--image-key' option
  - zdev: Allow dynamic control of module load
  - zipl/boot: Fix EBCDIC code page 500 conversion and decrease size by 200 bytes
  - zipl: Add support of heterogeneous mirrors (remove technical limitations
    on mirrored targets, thus allowing mirrored devices consist of partitions
    at different offsets on disks of different types and geometry).
  - zkey: Add support for generating and importing exportable secure keys

  Bug Fixes:
  - chpstat: Fix scaling of DPU utilization calculation
  - zdev/dracut: Prevent loading of unused kernel modules
  - zdev: Fix double device configuration on DPM systems
  - zdev: Fix double device configuration with rd.dasd
  - zipl_helper.device-mapper: Fix segfault in an error path

* __v2.38.0 (2025-06-25)__

  For Linux kernel version: 6.14 / 6.15

  Add new tools:
  - udev: New rule to set newly hotplugged CPUs online
  - zmemtopo: Display memory topology information
  - zpwr: Display power readings of a partition and CPC

  Removed tools / features:
  - check_hostkeydoc: Remove installation target
  - scsi_logging_level: Delete SCSI logging script (available in sg3_utils)
  - zdump: Drop build_arch for s390 DASD dumps
  - zdump: Drop non-extended multi-volume DASD dump support
  - zdump: Drop support of 32-bit dump architecture
  - zdump: Drop support of non-extended single volume DASD dumpers
  - zdump: Drop support of obsolete dumps and dumpers

  Changes of existing tools / libraries:
  - Various man-pages fixes
  - check_hostkeydoc: Add deprecation warning
  - check_hostkeydoc: Move to scripts directory
  - cpuplugd: Allow cpu hotplugging on systems without polarization
  - dbginfo.sh: Add Ubuntu snap tool
  - dbginfo.sh: Add missing config data and logs
  - dbginfo.sh: Reworking the container section
  - dbginfo.sh: Update for network commands
  - dbginfo.sh: Updating info for disks and lvm
  - libutil: Add machine type definition for machines 9175 and 9176
  - lscpumf: Add support for IBM z17 counter sets
  - lshwc: Add command line flag for run time
  - lshwc: Add flags to display counter values in hex
  - lshwc: Add output '--format' option

zconf/chp/lschp.c
/*
 * lschp - List information about available channel-paths
 *
 * Provide main function and command line parsing.
 *
 * Copyright IBM Corp. 2016, 2019
 *
 * s390-tools is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See LICENSE for details.
 */

#include <err.h>
#include <stdbool.h>
#include <stdlib.h>
#include <ctype.h>

#include "lib/util_base.h"
#include "lib/util_file.h"
#include "lib/util_fmt.h"
#include "lib/util_opt.h"
#include "lib/util_path.h"
#include "lib/util_prg.h"
#include "lib/util_rec.h"
#include "lib/util_scandir.h"
#include "lib/zt_common.h"

#include "lschp_cli.h"

#define CHP_SHARED  "shared"
#define CHP_STATE   "vary"
#define CHP_PCHID   "pchid"
#define CHP_TYPE    "type"
#define CHP_CFG     "cfg"
#define CHP_CMG     "cmg"
#define CHP_ID      "chpid"

static enum util_fmt_t fmt;
static bool fmt_specified;

/*
 * Program configuration
 */
static const struct util_prg prg = {
	.desc	= "List information about available channel-paths.",
	.copyright_vec = {
		{
			.owner = "IBM Corp.",
			.pub_first = 2016,
			.pub_last = 2019,
		},
		UTIL_PRG_COPYRIGHT_END
	}
};

/**
 * Set the "--format" option
 *
 * @param[in] str	Format for structured output
 */
static void fmt_set(const char *str)
{
	if (!util_fmt_name_to_type(str, &fmt)) {
		errx(EXIT_FAILURE, "Unknown format '%s', supported formats: %s\n", str,
		     FMT_TYPE_NAMES);
	}
}

/**
 * Print buffer structure as formatted
 *
 * @param[in] rec     The buffer structure, where results are written to
 */
static void fmt_rec_print_formatted(struct util_rec *rec)
{
	const char *str_val, *key;
	struct util_rec_fld *fld;
	bool is_chid_external;
	int ival = -1;

	util_fmt_obj_start(FMT_ROW, "entry");
	util_list_iterate(__util_rec_get_list(rec), fld) {
		key = util_rec_fld_get_key(fld);
		str_val = util_rec_get(rec, key);
		if (!strcmp(str_val, "-")) {
			util_fmt_pair(FMT_QUOTE | FMT_INVAL, key, "");
		} else if (!strcmp(key, CHP_PCHID)) {
			is_chid_external = (str_val[0] != '(');
			sscanf(str_val, is_chid_external ? "%x" : "(%x)", &ival);
			util_fmt_pair(FMT_DEFAULT, "chid", "%d", ival);
			util_fmt_pair(FMT_DEFAULT, "chid_external", "%s", is_chid_external ?
							"true" : "false");
		} else if (!strcmp(key, CHP_ID)) {
			util_fmt_pair(FMT_DEFAULT | FMT_QUOTE, key, "%s", str_val);
		} else if (!strcmp(key, CHP_TYPE)) {
			sscanf(str_val, "%x", &ival);
			util_fmt_pair(FMT_DEFAULT, key, "%d", ival);
		} else {
			ival = atoi(str_val);
			util_fmt_pair(FMT_DEFAULT, key, "%d", ival);
		}
	}
	util_fmt_obj_end(); /* entry */
}

static void fmt_init(void)
{
	unsigned int flags = FMT_HANDLEINT | FMT_KEEPINVAL;

	if (!fmt_specified)
		return;
	if (fmt == FMT_CSV)
		flags |= FMT_NOMETA | FMT_QUOTEALL;
	else
		flags |= FMT_DEFAULT;
	util_fmt_init(stdout, fmt, flags, 1);
}

static void fmt_end(void)
{
	if (!fmt_specified)
		return;
	util_fmt_exit();
}

/**
 * Read all attributes of a desired directory
 *
 * Which attributes are to be investigated is defined in the
 * fallowing function body.
 *
 * @param[in] dir     Path of the desired directory
 * @param[in] css_id  ID for device identification
 * @param[in] rec     The buffer structure, where results are written to
 * @param[in] chp     If set: CHPID to filter by
 */
static void print_chpid(const char *chp_dir, unsigned int css_id,
			struct util_rec *rec, char *chp)
{
	unsigned int css_id_tmp, chp_id;
	bool chid_external;
	char *path, buf[8];

	/* Get CHPID ID */
	if (sscanf(chp_dir, "chp%x.%x", &css_id_tmp, &chp_id) != 2)
		err(EXIT_FAILURE, "Invalid directory: %s", chp_dir);
	if (css_id != css_id_tmp)
		errx(EXIT_FAILURE, "Inconsistent css ids");
	path = util_path_sysfs("devices/css%x/%s", css_id, chp_dir);

	/* chpid  */
	util_rec_set(rec, CHP_ID, "%x.%02x", css_id, chp_id);

	/* vary */
	util_file_read_line(buf, sizeof(buf), "%s/status", path);
	if (strcmp(buf, "online") == 0)
		util_rec_set(rec, CHP_STATE, "1");
	else if (strcmp(buf, "offline") == 0)
		util_rec_set(rec, CHP_STATE, "0");
	else
		util_rec_set(rec, CHP_STATE, "-");

	/* configure */
	if (util_file_read_line(buf, sizeof(buf), "%s/configure", path))
		util_rec_set(rec, CHP_CFG, "-");
	else
		util_rec_set(rec, CHP_CFG, buf);

	/* type */
	if (util_file_read_line(buf, sizeof(buf), "%s/type", path))
		util_rec_set(rec, CHP_TYPE, "%s", "-");
	else
		util_rec_set(rec, CHP_TYPE, "%02lx", strtoul(buf, NULL, 16));

	/* cmg */
	util_file_read_line(buf, sizeof(buf), "%s/cmg", path);
	if ((strcmp(buf, "unknown") == 0) || (strlen(buf) == 0))
		util_rec_set(rec, CHP_CMG, "%s", "-");
	else
		util_rec_set(rec, CHP_CMG, "%-3lx", strtoul(buf, NULL, 0));

	/* shared */
	util_file_read_line(buf, sizeof(buf), "%s/shared", path);
	if ((strcmp(buf, "unknown") == 0) || (strlen(buf) == 0))
		util_rec_set(rec, CHP_SHARED, "%s", "-");
	else
		util_rec_set(rec, CHP_SHARED, "%-6lx", strtoul(buf, NULL, 0));

	/* chid */
	util_file_read_line(buf, sizeof(buf), "%s/chid_external", path);
	if (strcmp(buf, "1") == 0)
		chid_external = true;
	else
		chid_external = false;
	util_file_read_line(buf, sizeof(buf), "%s/chid", path);
	if (strlen(buf) != 0) {
		if (chid_external)
			util_rec_set(rec, CHP_PCHID, " %4s ", buf);
		else
			util_rec_set(rec, CHP_PCHID, "(%4s)", buf);
	} else {
		util_rec_set(rec, CHP_PCHID, "%s", "-");
	}
	if (!strlen(chp) || strcmp(util_rec_get(rec, CHP_ID), chp) == 0) {
		if (fmt_specified)
			fmt_rec_print_formatted(rec);
		else
			util_rec_print(rec);
	}
	free(path);
}

/**
 * Compare two dirents numerically (chp0.ff <-> chp0.fe)
 *
 * @param[in] de1   First directory entry
 * @param[in] de2   Second directory entry
 * @retval    -1    de1 is lower
 *             0    de1 equals de2
 *             1    de2 is lower
 */
static int chpsort(const struct dirent **de1, const struct dirent **de2)
{
	unsigned long val1 = strtoul(&(*de1)->d_name[5], NULL, 16);
	unsigned long val2 = strtoul(&(*de2)->d_name[5], NULL, 16);

	if (val1 < val2)
		return -1;
	if (val1 == val2)
		return 0;
	return 1;
}

/**
 * Show all channel paths for a given directory in sysfs
 *
 * @param[in] css_dir The desired directory
 * @param[in] rec     The buffer structure, where results are written to
 * @param[in] chp     If set: CHPID to filter by
 */
static void print_css(const char *css_dir, struct util_rec *rec, char *chp)
{
	struct dirent **de_vec;
	unsigned int css_id;
	int i, count;
	char *path;

	if (sscanf(css_dir, "css%x", &css_id) != 1)
		err(EXIT_FAILURE, "Invalid directory: %s", css_dir);

	path = util_path_sysfs("devices/css%d", css_id);
	count = util_scandir(&de_vec, chpsort, path, "chp%x.*", css_id);
	for (i = 0; i < count; i++)
		print_chpid(de_vec[i]->d_name, css_id, rec, chp);
	util_scandir_free(de_vec, count);
	free(path);
}

/*
 * Print chpid table
 *
 * @param[in] chp              If set: CHPID to filter by
 */
static void cmd_lschp(char *chp)
{
	struct dirent **de_vec;
	struct util_rec *rec;
	int i, count;
	char *path;

	fmt_init();
	rec = util_rec_new_wide("=");
	util_rec_def(rec, CHP_ID,	UTIL_REC_ALIGN_LEFT, 6, "CHPID");
	util_rec_def(rec, CHP_STATE,	UTIL_REC_ALIGN_LEFT, 5, "Vary");
	util_rec_def(rec, CHP_CFG,	UTIL_REC_ALIGN_LEFT, 5, "Cfg.");
	util_rec_def(rec, CHP_TYPE,	UTIL_REC_ALIGN_LEFT, 5, "Type");
	util_rec_def(rec, CHP_CMG,	UTIL_REC_ALIGN_LEFT, 4, "Cmg");
	util_rec_def(rec, CHP_SHARED,	UTIL_REC_ALIGN_LEFT, 6, "Shared");
	util_rec_def(rec, CHP_PCHID,	UTIL_REC_ALIGN_LEFT, 6, " PCHID");

	if (!fmt_specified)
		util_rec_print_hdr(rec);
	/*
	 * Iterate over each "/sys/devices/css.*"
	 */
	path = util_path_sysfs("devices");
	count = util_scandir(&de_vec, alphasort, path, "^css[[:xdigit:]]{1,2}$");
	if (fmt_specified)
		util_fmt_obj_start(FMT_LIST, "channel_paths");
	for (i = 0; i < count; i++)
		print_css(de_vec[i]->d_name, rec, chp);
	if (fmt_specified)
		util_fmt_obj_end(); /* channel_paths */
	util_ptr_vec_free((void **) de_vec, count);
	free(path);
	util_rec_free(rec);
	fmt_end();
}

#define CHP_LEN		4
/*
 * Parse options and execute the command
 */
int main(int argc, char *argv[])
{
	char chp[CHP_LEN + 1] = "";
	int c;

	util_prg_init(&prg);
	util_opt_init(lschp_opt_vec, NULL);

	while (1) {
		c = util_opt_getopt_long(argc, argv);
		if (c == -1)
			break;
		switch (c) {
		case 'h':
			util_prg_print_help();
			util_opt_print_help();
			return EXIT_SUCCESS;
		case 'v':
			util_prg_print_version();
			return EXIT_SUCCESS;
		case OPT_FORMAT:
			fmt_specified = true;
			fmt_set(argv[optind - 1]);
			break;
		default:
			util_opt_print_parse_error(c, argv);
			return EXIT_FAILURE;
		}
	}
	if (argc > optind + 1)
		errx(EXIT_FAILURE, "Too many arguments specified");
	if (argc == optind + 1) {
		/* we take a single argument only */
		switch (strlen(argv[optind])) {
		case 1:
			sprintf(chp, "0.0%s", argv[optind]);
			break;
		case 2:
			sprintf(chp, "0.%s", argv[optind]);
			break;
		case CHP_LEN:
			strcpy(chp, argv[optind]);
			break;
		default:
			errx(EXIT_FAILURE, "%s is not a valid channel-path ID",
			     argv[optind]);
		}
		if (!isdigit(chp[0]) || chp[1] != '.' || !isxdigit(chp[2]) ||
		    !isxdigit(chp[3]))
			errx(EXIT_FAILURE, "%s is not a valid channel-path ID",
			     chp);
	}
	cmd_lschp(chp);

	return EXIT_SUCCESS;
}
